src/osp/id_registry.h

```diff
@@ -0,0 +1,341 @@
+/**
+ * Open Space Program
+ * Copyright © 2019-2021 Open Space Program Project
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#pragma once
+
+#include "hierarchical_bitset.h"
+
+#include <cassert>
+#include <limits>
+#include <utility>
+#include <vector>
+
+namespace osp
+{
+
+template<class TYPE_T>
+constexpr TYPE_T id_null() noexcept
+{
+    using underlying_t = typename std::underlying_type_t<TYPE_T>;
```

Comment:
careful if TYPE_T isn't an enum. std::underlying_type_t, doing exactly the opposite of what everyone would expect, will cause a compile error if given "int".

Recommend adding static_assert(std::is_enum_v<TYPE_T>) right above this line so that a *very clear* error message happens, instead of the template garbage that you'll get from std::underlying_type_t

Reply:
Huh. I'd still like regular integers to work however. How about something like:
```cpp
using underlying_t = std::conditional_t<std::is_enum_v<TYPE_T>, std::underlying_type_t<TYPE_T>, TYPE_T>;
static_assert(std::is_integral_v<underlying_t>);
```